scripts/ubiquiti-discovery.nse

```diff
@@ -0,0 +1,254 @@
+-- <elem key="product">ER-X</elem>
+-- <elem key="firmware">EdgeRouter.ER-e50.v1.10.7.5127989.181001.1227</elem>
+-- <elem key="version">v1.10.7</elem>
+-- <table key="mac_ip">
+--   <elem key="80:2a:a8:df:a1:63">192.168.0.1</elem>
+--   <elem key="80:2a:a8:df:a1:5e">55.55.55.55</elem>
+-- </table>
+-- <table key="mac_addresses">
+--   <elem>80:2a:a8:df:a1:63</elem>
+--   <elem>80:2a:a8:df:a1:5e</elem>
+-- </table>
+
+
+portrule = shortport.port_or_service(10001, "ubiquiti-discovery", "udp", {"open", "open|filtered"})
+
+PROBE_V1 = string.pack("BB I2",
+  0x01, 0x00, -- version, command
+  0x00, 0x00  -- length
+)
+
+---
+-- Converts uptime seconds into a human readable string
+--
+-- E.g. "86518" -> "1 days 00:01:58"
+--
+-- @param uptime number of seconds of uptime
+-- @return formatted uptime string (days, hours, minutes, seconds)
+local function uptime_str(uptime)
+  local uptime_num = tonumber(uptime)
+  if not uptime_num then
+    return nil
+  end
+
+  local d = math.floor(uptime_num / 86400)
+  local h = math.floor(uptime_num /  3600 % 24)
+  local m = math.floor(uptime_num /    60 % 60)
+  local s = math.floor(uptime_num % 60)
+
+  return string.format("%d days %02d:%02d:%02d", d, h, m, s)
+end
+
+---
+-- Parses the full payload of a discovery response
+--
+-- @param payload containing response
+-- @return output_table containing results or nil
+local function parse_v1_discovery_response(response)
+
+  local info = stdnse.output_table()
+  local unique_macs = {}
+  local mac_ip_table = {}
+
+  if #response < 4 then
+    return nil
+  end
+
+  -- Check for v1 protocol header
+  if not ( response:byte(1) == 0x01 and response:byte(2) == 0x00 ) then
+    return nil
+  end
+
+  local config_len = string.unpack(">I2", response, 3)
+
+  -- Do the lengths check out?
+  if ( not ( #response == config_len + 4) ) then
+    return nil
+  end
+
+  -- Response looks legit, start extraction
+  local config_data = string.sub(response, 5, #response)
+
+  local tlv_type, tlv_len, tlv_value, pos
+  local mac, mac_raw, ip, ip_raw
+  pos = 1
+
+  while pos <= #config_data - 2 do
+    tlv_type = config_data:byte(pos)
+    tlv_len  = string.unpack(">I2", config_data, pos +1)
+    pos = pos + 3
+
+    -- Sanity check that TLV len isn't larger than the data we have left.
+    -- Has been observed in the wild against protocols just similar enough to
+    -- make it here.
+    if tlv_len > (#config_data - pos + 1) then
+      return nil
+    end
+
+    tlv_value = config_data:sub(pos, pos + tlv_len - 1)
+
+    -- MAC address
+    if tlv_type == 0x01 then
+      mac_raw = tlv_value:sub(1, 6)
+      mac = stdnse.format_mac(mac_raw)
+      unique_macs[mac] = true
+
+    -- MAC and IP address
+    elseif tlv_type == 0x02 then
+      mac_raw = tlv_value:sub(1, 6)
+      mac = stdnse.format_mac(mac_raw)
+      unique_macs[mac] = true
+
+      ip_raw = tlv_value:sub(7, tlv_len)
+      ip = ipOps.str_to_ip(ip_raw)
+      mac_ip_table[mac] = ip
```

Comment:
Is it possible that a single MAC could be associated with more than one IP address? Maybe IPv4 and IPv6?